core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+  }
+
+  def extractConfigFromKafkaConfig(kafkaConfig: KafkaConfig): Config = Config(kafkaConfig.longTailProduceRequestLogThresholdMs,
+                                                                              kafkaConfig.longTailProduceRequestLogRatio)
+
+  /**
+   * Calculate the time in each stage by
+   * 1. Sort stages by timestamp
+   * 2. Make a sliding window of 2 on 1.
+   * 3. Diff the timestamp on the window to get the time spent on the stage of the first element in 2
+   */
+  def toTimeTakenInEachStageMessage(instrumentation: ProduceRequestInstrumentation): String = {
+    val marks = instrumentation.marks
+
+    // Enforce Finish is marked, so when sliding there is at list a full window
+    marks.get(Stage.Finish) match {
+      case None => instrumentation.markStage(Stage.Finish)
+      case _ =>
+    }
+
+    marks
+      .toSeq                          // Iterate over stages
+      .sortBy(s => (s._2, s._1.id))   // in order of marked timestamp then defined order of stage (to prevent multiple stages finish within 1ms)
+      .sliding(2)                     // Make sliding window to diff time taken in each stage
+      .map { window =>
+        val iter = window.iterator
+        val (stage1, timestamp1) = iter.next()
+        val (_, timestamp2) = iter.next()
+        // Use the next stage's timestamp mark minus this stage's timestamp mark to get time spent on current stage
+        // Last stage is on finish, no diff to check, i.e. no need to log
+        s"${stage1.toString}: ${timestamp2 - timestamp1}"
+      }
+      .mkString("(", ", ", ")")
+  }
+
+  /**
+   * Based on the config (e.g. probability funnel), logs the detail of the instrumentation
+   *
+   * @param request This is required to obtain the total time for threshold funnel
+   * @param instrumentation The instrumentation to log
+   */
+  def maybeLog(request: RequestChannel.Request,
+               instrumentation: ProduceRequestInstrumentation): Unit = {
+    val totalTimeMs = nanosToMs(time.nanoseconds() - request.startTimeNanos)
+    if (totalTimeMs < config.thresholdToLogMs) {
+      logger.debug("totalTimeMs={} is below {}={}, skip logging",
+                   totalTimeMs,
+                   KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+                   config.thresholdToLogMs)
+      return
+    }
+
+    val coin = rnd.nextDouble()
+    if (coin > config.logRatio) {
+      logger.debug("Funneled out by random coin {} > configured ratio {}", coin, config.logRatio)
+      return
+    }
+
+    // Details like instrumentation.topicPartitionsInRequest cannot be obtained from request object,
+    // because the underlying details, at this moment, have been cleared by ProduceRequest#clearPartitionRecords()
+    logger.info(
+      s"totalTimeMs=${totalTimeMs}; acks=${instrumentation.requiredAcks}; "
+        + s"topic_partitions=${instrumentation.topicPartitionsInRequest.map(tp => s"${tp.topic()}-${tp.partition()}").mkString("(", ", ", ")")}; "
```

Comment:
I feel information of topic partitions is not important here unless the identity of partition affects the latency. If the number of partitions are big, this information would spam in the log.

Reply:
I was thinking of that too.  But it could be that this is actually related (e.g. what if a large partition info covered is critical cause to the long-tailed event).

Reply:
I agree, we're quite likely to be debugging a particular slow produce request, and that info might also be needed to find the proper log line.